### PR TITLE
Add Permit

### PR DESCRIPTION
**Pull request by banteg**

### CODE REVIEW

File: contracts/Cornichon.vy [hunk 1]
```diff
@@ -23,21 +23,43 @@ name: public(String[64])
 symbol: public(String[32])
 decimals: public(uint256)
 balanceOf: public(HashMap[address, uint256])
+nonces: public(HashMap[address, uint256])
 allowances: HashMap[address, HashMap[address, uint256]]
+version: public(String[32])
```
Comment (fubuloubu) [inline contracts/Cornichon.vy]: Make Version a string constant, and have a view-only method that returns the constant to save gas

File: contracts/Cornichon.vy [hunk 2]
```diff
@@ -23,21 +23,43 @@ name: public(String[64])
...
+            convert(chain.id, bytes32),
+            convert(self, bytes32)
+        )
+    )
+
+
+@view
+@external
+def dom_sep() -> uint256:
+    return chain.id
```
Comment (fubuloubu) [inline contracts/Cornichon.vy]: Is this necessary? `eth_chainId` is an RPC method that's available.

File: tests/test_permit.py [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+            "spender": spender.address,
+            "amount": amount,
+            "nonce": nonce,
+            "expiry": expiry,
+        },
+    }
+    message = encode_structured_data(data)
+    signed = owner.sign_message(message)
+    assert encode_hex(hash_domain(data)) == cornichon.DOMAIN_SEPARATOR()
+    tx = cornichon.permit(owner.address, spender.address, amount, nonce, expiry, signed.signature)
```
Comment (fubuloubu) [inline tests/test_permit.py]: why is `tx` here? for debugging?

File: tests/test_permit.py [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+    assert cornichon.allowance(owner.address, spender.address) == amount
```
Comment (fubuloubu) [inline tests/test_permit.py:52]: add that it's 0 to start above the `permit` action

**Review by fubuloubu — APPROVED** [review]

File: contracts/Cornichon.vy (hunk 2 above)
Comment (banteg) [reply contracts/Cornichon.vy]: Left this form testing, removed now

File: contracts/Cornichon.vy (hunk 1 above)
Comment (banteg) [reply contracts/Cornichon.vy]: Should I rather remove it altogether and make the domain separator just (chain.id, self)?

**Pull request merged**